mitmproxy/addons/dns_resolver.py

```diff
@@ -40,25 +39,65 @@ def configure(self, updated):
             self.resolver.cache_clear()
             self.name_servers.cache_clear()
 
-    @cache
-    def resolver(self) -> mitmproxy_rs.DnsResolver:
-        return mitmproxy_rs.DnsResolver(
-            name_servers=self.name_servers(),
-            use_hosts_file=ctx.options.dns_use_hosts_file,
-        )
-
     @cache
     def name_servers(self) -> list[str]:
+        """
+        The Operating System name servers,
+        or `None` if they cannot be determined.
+        """
         try:
             return ctx.options.dns_name_servers or mitmproxy_rs.get_system_dns_servers()
         except RuntimeError as e:
-            raise RuntimeError(
-                f"Failed to get system dns servers: {e}\nMust set dns_name_servers option to run DNS mode."
+            logger.warning(
+                f"Failed to get system dns servers: {e}\n"
+                f"The dns_name_servers option needs to be set manually."
             )
+            return []
+
+    @cache
+    def resolver(self) -> mitmproxy_rs.DnsResolver:
+        """
+        Our mitmproxy_rs DNS resolver.
+        """
+        ns = self.name_servers()
+        assert ns
+        return mitmproxy_rs.DnsResolver(
+            name_servers=ns,
+            use_hosts_file=ctx.options.dns_use_hosts_file,
+        )
 
     async def dns_request(self, flow: dns.DNSFlow) -> None:
-        assert flow.request
-        should_resolve = (
+        if self._should_resolve(flow):
+            all_ip_lookups = (
+                flow.request.query
+                and flow.request.op_code == dns.op_codes.QUERY
+                and flow.request.question
+                and flow.request.question.class_ == dns.classes.IN
+                and flow.request.question.type in (dns.types.A, dns.types.AAAA)
+            )
+            name_servers = self.name_servers()
+
+            if all_ip_lookups:
+                # For A/AAAA records, we try to use our own resolver
```

Comment:
I think using our own resolver here is slightly advantageous because it tries multiple DNS servers. If we just forward to `name_servers[0]` and that one is offline, we're in a bad spot. 🙃